Add RESTful services spec

kahowell opened the pull request:
This is intended to cover both data and non-data services, and provide
some consistency between the two.

Please take a look; poke holes/discuss/disagree. Feel free to create other PRs based off this one, etc.

If anyone cares strongly enough, I can move this to a separate repo.

Tagging: @snehagunta @BVulaj @jewzaam @dcrissman @bserdar @paterczm @derek63 @alechenninger @bdmiller3 @macgregor @kahoali

Please tag others if you'd like.

kahowell committed:
Add RESTful services spec

This is intended to cover both data and non-data services, and provide
some consistency between the two.

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+Endpoints may not support all methods (i.e. a service may provide read-only
+access to a resource, or a service may not support certain operations for
+performance or other considerations; for example, `GET` may not be supported
+for `user` if listing all users is disallowed for performance reasons.
```
Maybe worth calling out idempotency/side effect requirements of HTTP methods. ~~Mainly because when it comes to RPC type API's, sometimes it may not be obvious what verb to use, and idempotency requirements can maybe make the decision easy.~~

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+## URIs
+
+Version granularity is up to implementation. However, the base URI for an
+entity must include a version.
```
Is it bike-shedding to standardize on a single major version number representing backwards compatibility?

Personally I think it's pretty important but I could be swayed. To summarize:
- I think point of a version number is so that you can support multiple incompatible versions at one time. Otherwise, the version number will never change, so there is no point to it.
- If you're going to support multiple at one time, that's hard, so you might want to reduce the amount of simultaneously deployed versions to a minimum: the latest release of each major version. Assuming you have some sane deprecation plan in place and your rate of breaking changes is low, you probably restrict yourself to one or two versions at a time with this rule. Otherwise you could have a lot of minor version variations supported which could be a maintainability problem that would could potentially get rid of wholesale by restricting to a single version (ex: v1.1, v1.2, v2.0, v2.1 vs v1 and v2).

kahowell replied:
By "version granularity", I really meant whether the whole set of services gets versioned, vs. a set of services for a given entity. Open to suggestions on how to express this. Or we could just make it explicit for the sake of avoiding bike-shedding such as:
- The APIs shall be versioned per entity.

I agree with your points. Let's start with an actual standard and adjust course if necessary... How about the following?
- The APIs shall be versioned with a major version only. APIs may be added without incrementing the version, but they must not be removed or changed in breaking ways (possibly reference semver for minor/patch semantics?).
- Multiple versions of the API may be supported.

If I don't get disagreement in the next couple of days, I'll go ahead and make these changes :-)

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+Specific resources will be addressed through the base URI plus the identifier
+for that resource.
+
+    /user/{id}
```
This implies that all resources have a single id. We have some compound ids.

Personally I've leaned towards not supporting path queries like this and requiring the use of [queries](https://github.com/lightblue-platform/lightblue-docs-specs/pull/43/files#r66460928) always, as unfortunately verbose as this may be. 

Another option would be to require that every resource has a single id, but this might be more controversial. I like the simplicity of that, but sometimes it forces you to leave the natural domain/semantics of the resource (imagine a case where business users always talk about identity of a domain object in terms of two values, but now we require it's determined by an id, which means we have 2 ways of getting at the same thing, etc... seems maybe messy).

If there were simple query param based queries it might be moot (in that case you just do /user?id={id}&otherKey={otherKey}).

What do you think?

kahowell replied:
Good point. I don't like the complexity of parsing the URIs that using a scheme such as `user?id={id}&otherKey={otherKey}` introduces... (consider especially operations against a specific instance) So I'd rather encode the compound key some other way... How about this?
- If the entity has an identity which involves multiple fields, then the `{id}` shall consist of each field separated by `_` in a consistent, entity-specific order.

I don't think that covers _all_ possibilities, but I think as long as it covers current needs, it's fine to make an exception/new rule if the need arises...

alechenninger replied:
Hrm yeah bummer query params must be at the end of URL. Personally I'm not crazy about the delimiter-based syntax. I'm concerned of scenarios where the order would be ambiguous and a user would have to look this up. Plus, I've never seen this in the wild so it's a surprising convention. And maybe the delimiter would have to be escaped in rare cases. I could be swayed, however. Maybe with a different delimiter.

But, consider for operations, maybe we don't require /resource/123/operation support, and instead a user can do /resource/operation, with all parameters in the POST body. To some extent this reduces the surface area of the spec since /resource/operation is there either way.

This might be a good scenario to sort of weigh out pro/cons of each approach.

bvulaj replied:
I've seen the delimited version in the wild, but I also have mixed feelings on it.  Though, I have mixed feelings on the query param version as well.  I think we're kicking the discoverable portion HATEOAS principles to the curb by going that route.

bvulaj replied:
HAL has an idea of `templated` URIs, but they only serve as examples of how to use query params, not as a discoverable list of resources.

bvulaj replied:
https://library.oreilly.com/book/9780596801694/restful-web-services-cookbook/38.xhtml

Adding for reference.

```
http://www.example.org/co-ordinates;w=39.001409,z=-84.578201
http://www.example.org/axis;x=0,y=9 
```

dewayneHat replied:
> Another option would be to require that every resource has a single id, but this might be more controversial. I like the simplicity of that, but sometimes it forces you to leave the natural domain/semantics of the resource (imagine a case where business users always talk about identity of a domain object in terms of two values, but now we require it's determined by an id, which means we have 2 ways of getting at the same thing, etc... seems maybe messy).

IMO, APIs are not for business users, they are for developers and the services/applications they write.  If using a single id to identify resources that have compound ids makes the API more useable for the developer writing code against it, then that is what we should do.  The business can continue to think of it their way without knowing the implementation details of the API.

/users/{id}
I think single id is probably the api that will end up being the most intuitive and natural to use as a consumer of the API. However, this could lead to situations where a consumer of the API may have the compound id of the resource, and not the "api id".  So we might end up creating an api to support that as well anyhow....

jewzaam replied:
We have had problems with path params in the past, specifically that path
params are not encoded automatically by libraries as they are for query
params.  So I see some risk there.  Also, with a path param you have to
pick the order for those parameters.  What is intuitive to you, is not
intuitive to me.  Using a query param means order doesn't matter and I have
a more consistent experience.  How would your example account for searching
for a user by login?  What if my login was a number and that number was
also someone's ID?  Ambiguity is a problem that goes away with query
params, you have to explicitly identify with a parameter name all the
values you want in your request.

On Fri, Jun 24, 2016 at 8:58 AM, dewayneHat notifications@github.com
wrote:

> In rest_specification/restful_service_specification.md
> https://github.com/lightblue-platform/lightblue-docs-specs/pull/43#discussion_r68393921
> :
> 
> > +| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
> > +| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
> > +| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
> > +| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
> > +| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
> > +
> > +Endpoints may not support all methods (i.e. a service may provide read-only
> > +access to a resource, or a service may not support certain operations for
> > +performance or other considerations; for example, `GET` may not be supported
> > +for `user` if listing all users is disallowed for performance reasons.
> > +
> > +## Operations Against Specific Resources
> > +Specific resources will be addressed through the base URI plus the identifier
> > +for that resource.
> > +
> > -    /user/{id}
> 
> Another option would be to require that every resource has a single id,
> but this might be more controversial. I like the simplicity of that, but
> sometimes it forces you to leave the natural domain/semantics of the
> resource (imagine a case where business users always talk about identity of
> a domain object in terms of two values, but now we require it's determined
> by an id, which means we have 2 ways of getting at the same thing, etc...
> seems maybe messy).
> 
> IMO, APIs are not for business users, they are for developers and the
> services/applications they write. If using a single id to identify
> resources that have compound ids makes the API more useable for the
> developer writing code against it, then that is what we should do. The
> business can continue to think of it their way without knowing the
> implementation details of the API.
> 
> /users/{id}
> I think single id is probably the api that will end up being the most
> intuitive and natural to use as a consumer of the API. However, this could
> lead to situations where a consumer of the API may have the compound id of
> the resource, and not the "api id". So we might end up creating an api to
> support that as well anyhow.
> 
> For the sake of example let's say we recently merged users from several
> different databases into a single db. And for some reason, we decided to
> make the user's compound id: 'id' and 'dborigin'. Where dborigin is the
> database where the user originated from (0=Oracle, 1=mysql, 2=mongo). Then
> the api could look something like this
> GET /users -- Returns all users (Paginated of course)
> GET /users/1 -- Returns all users with id 1 (Max of 3)
> GET /users/1/dborigin/0 -- Returns single user with id 1 from Oracle db
> 
> I know this isn't perfect, but it makes sense intuitively.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/lightblue-platform/lightblue-docs-specs/pull/43/files/19a07e2017455997f1b9c4bdb15edb6e3cceb8ce#r68393921,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAdUhJGaWWp4MvkrO-W6pX-3rnheUXDDks5qO9RngaJpZM4IopeE
> .

bvulaj replied:
Query params are for filtering.  Path params are for navigation and reference.  Let's not confuse them.

HATEOAS and REST principles have embraced single IDs.  So what if some object has a unique constraint on X fields?  That does not have to be the REST identifier.  IMO, it's going to get messy if we don't enforce that in the spec.

Examples:

Single User URI:
`/users/1`
Search for User by Login
`/user/search?login=foo`
Get a User by their "business identifier":
`/user/search?x=foo&y=bar`

dewayneHat replied:
> /user/search?login=foo
> +1 for query params when filtering content. 

To echo @BVulaj 
`GET path/{id}` makes identification of the resource more uniform and self descriptive to the consumer with little knowledge of the server code. 

`GET path?x=foo&y=bar&z=car`. For **every** resource that is navigable via query params _only_, the consumer has to determine for each combination of fields: Am I specifying search criteria, identifying a single resource, or multiple resources?  Can I PUT the resource(s) I just received back to that same endpoint? We would need really strong documentation in order for a consumer to use the API properly. 

Where as `GET /users/id` and `GET /users/search?x=foo&y=bar&z=car` tell me more about the action I'm asking the server to perform. I know intuitively, that I can't PUT back on `/users/search?...` results. It doesn't make sense to try, and if I do, the server will save me and I should get a 404. Where as the `PUT /users?.....` might actually try to make that happen, maybe with unexpected results if I pass the wrong combination of fields.

bserdar replied:
Something like this might make more sense:

 GET path/id -> retrieve by id, only for simple identities, won't work for
composite id
 POST path/search?<query> -> search operation

This is following the convention that GET without a verb means retrieve
entity with id, and POST with a verb means perform an operation.

On Fri, Jun 24, 2016 at 1:41 PM, dewayneHat notifications@github.com
wrote:

> In rest_specification/restful_service_specification.md
> https://github.com/lightblue-platform/lightblue-docs-specs/pull/43#discussion_r68451929
> :
> 
> > +| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
> > +| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
> > +| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
> > +| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
> > +| `DELETE`       | Delete ALL resources   | -                             | -            | -              |
> > +
> > +Endpoints may not support all methods (i.e. a service may provide read-only
> > +access to a resource, or a service may not support certain operations for
> > +performance or other considerations; for example, `GET` may not be supported
> > +for `user` if listing all users is disallowed for performance reasons.
> > +
> > +## Operations Against Specific Resources
> > +Specific resources will be addressed through the base URI plus the identifier
> > +for that resource.
> > +
> > -    /user/{id}
> 
> /user/search?login=foo
> +1 for query params when filtering content.
> 
> To echo @BVulaj https://github.com/BVulaj
> GET path/{id} makes identification of the resource more uniform and self
> descriptive to the consumer with little knowledge of the server code.
> 
> GET path?x=foo&y=bar&z=car. For _every_ resource that is navigable via
> query params _only_, the consumer has to determine for each combination
> of fields: Am I specifying search criteria, identifying a single resource,
> or multiple resources? Can I PUT the resource(s) I just received back to
> that same endpoint? We would need really strong documentation in order for
> a consumer to use the API properly.
> 
> Where as GET /users/id and GET /users/search?x=foo&y=bar&z=car tell me
> more about the action I'm asking the server to perform. I know intuitively,
> that I can't PUT back on /users/search?... results. It doesn't make sense
> to try, and if I do, the server will save me and I should get a 404. Where
> as the PUT /users?..... might actually try to make that happen, maybe
> with unexpected results if I pass the wrong combination of fields.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/lightblue-platform/lightblue-docs-specs/pull/43/files/19a07e2017455997f1b9c4bdb15edb6e3cceb8ce#r68451929,
> or mute the thread
> https://github.com/notifications/unsubscribe/ADgDDV0TRZPbqnp8GI6118QILg4VrHztks5qPDL-gaJpZM4IopeE
> .

alechenninger replied:
First off, loving the discussion!

> Query params are for filtering. Path params are for navigation and reference.

Unfortunately, I don't think we can rely on these kinds of absolutes. We have to make compromises. REST implementations widely vary in practice. This is unfortunate, but c'est la vie.

> IMO, APIs are not for business users, they are for developers and the services/applications they write.

I think this line of reasoning can be risky long term. It is good to strive for [ubiquitous language](http://martinfowler.com/bliki/UbiquitousLanguage.html) between developers, QA, and business users. If we speak the same language, there is less lost in translation. The identifiers for a domain "thing" can be important, so I think it's not unreasonable to ask everyone to speak the same language here, though maybe not critical for ids. I would empathize with QA here particularly, though.

> GET path?x=foo&y=bar&z=car. For every resource that is navigable via query params only, the consumer has to determine for each combination of fields: Am I specifying search criteria, identifying a single resource, or multiple resources?

I think this is a good point, but I think maybe solving is not so bad. How about this: the answer is you are always identifying a single resource just like with a path param. We have in the spec that requests against /resource are for everything in that resource. We have in the spec that /resource/{function} can do anything. So that means anything else is one entity and one entity only, just like /user/{id} is. If not enough params are provided for a compound identity then a 4xx error response is returned according to error schema (below). GET/PUT/PATCH/DELETE should work as expected: one entity or fail. There is no ambiguity and the spec is consistent.

As far as having to know what query params to use, I agree it is definitely easier when everything is just a single path param id. Taking inspiration from @bserdar 's suggestion, maybe /user/{id} could be sugar for /user?id={id}. Then, if "id" is a param, it just works. If it isn't, you get a descriptive error that says what params need to be provided. What do you think?

Additionally, as far as consumer discoverability goes, I hope that whatever solution we land on has a clear route for good, auto-generated clients in many languages. This makes some of the details of the REST interface not as important to get perfect since most would see it through a language-specific client anyway.

My 2c. Good discussion!

Edit: Took a shot at updating the proposal here https://github.com/kahowell/lightblue-docs-specs/pull/1

dewayneHat replied:
>  We have in the spec that /resource/{function} can do anything. So that means anything else is one entity and one entity only, just like /user/{id} is. If not enough params are provided for a compound identity then a 4xx error response is returned according to error schema (below). GET/PUT/PATCH/DELETE should work as expected: one entity or fail. There is no ambiguity and the spec is consistent.

I think this approach is good enough.  My inner zealot is unhappy, however there probably are _some/enough/many_ cases in our domain set that won't work well with path/{id}. I just don't know enough about the domain, so I'll defer to the folks who have the domain expertise. 

Also agree, it is a good discussion. 😄

bvulaj replied:
@alechenninger Ask and ye shall receive.  https://github.com/mikekelly/hal_specification/wiki/Libraries

HAL is a great spec to work off of.  Though if we push the use of the clients, we do have to be more careful about single item resource navigation.

kahowell replied:
This might sound silly, but if a resource doesn't naturally have a single id field, we could generate a UUID for the resource, and use that as the REST id. Then to work against that resource, we'd have the user do /search?foo=bar&foo2=bar2 to retrieve the entity at first, and then use the UUID in subsequent operations... That would work around _any_ case where the natural identifier for a resource is made up of more than one field.

bvulaj replied:
It might sound silly, but who cares.  I think that could work if we can't guarantee an singular identifier otherwise.  The ID doesn't need to be human-understandable.  It's just a reference to a unique entity.  That doesn't mean the entity still can't have unique or compound constraints.  In what other contexts does our initial operation on an entity start with us punching in the identifier?  It's always going to be through some sort of search, followed by subsequent operations with the actual ID.

Maybe going down the rabbit hole here, but it's worth noting that in typical HAL and many popular implementations of it, there is no `id` field included in the response... the ID is the URI of the entity, since that's what HATEOAS operates on.

dcrissman replied:
Would the UUID need to be permanently persisted in this case?

bvulaj replied:
I don't think it needs to.  It could though, and would be helpful.

To double down on what @dewayneHat said - I'm not really privy to the entire domain and its requirements, so I'm just trying to be the best (worst?) REST / HATEOAS champion I can here, and hopefully in the end we all agree on something in the middle.

dewayneHat replied:
I think what you guys are saying is great. In fact, that'd be my preference @BVulaj @kahowell. But I would like to make sure that the design addresses or at least acknowledges as a weakness all issues on the table.  

Currently it doesn't address the following two issues.  
**Encoding issues**

> We have had problems with path params in the past, specifically that path
> params are not encoded automatically by libraries as they are for query
> params.

I'd like to hear more detail on @jewzaam.  I feel like having consistent numerical identifiers or UUIDs would be fine for this.

**Translation issues**

> I think this line of reasoning can be risky long term. It is good to strive for ubiquitous language between developers, QA, and business users. If we speak the same language, there is less lost in translation. The identifiers for a domain "thing" can be important, so I think it's not unreasonable to ask everyone to speak the same language here, though maybe not critical for ids. I would empathize with QA here particularly, though.

Although I still believe that business users shouldn't be concerned about API implementation, I do understand that such translation failures can lead to requirements replete with metric mixup mines.  I think the business users, if they truly are using/writing requirements for the API, will get used to having a different identifier used in APIs.

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+## Sorting
+
+Sorting shall be provided via the `sort` query parameter. `sort` may be
+specified more than once in order to sort by more than one field. `sort` values
```
Are you saying for multiple values, do this: `/user/1?sort=foo,desc&sort=bar,asc`?

kahowell replied:
Yes, this is a convention borrowed from Spring Data Rest (ex. http://stackoverflow.com/a/33034533). Maybe an example would help clarity though?

TODO:
- [ ] Add example of sort with multiple values

alechenninger replied:
That's a bit surprising. I wonder how widely supported multiple params are while retaining order? Is Java enough for us? From [this](https://www.owasp.org/images/b/ba/AppsecEU09_CarettoniDiPaola_v0.8.pdf) (page 9) it looks like many frameworks do different things when they encounter the same param multiple times, and from [this](http://stackoverflow.com/questions/1746507/authoritative-position-of-duplicate-http-get-query-keys) it sounds like there is no spec to be the decider. I'm getting the impression it would be safer to maybe do something like `sort=foo:asc,bar:desc` (specific syntax aside). Although, if we only care about Java this is probably moot, but personally I was hoping this could be useful beyond Java server apps (which is not to say that multiple of same param excludes them, but I guess it might get complicated).

kahowell replied:
Good point, I did a quick survey to see how support for this is:
- Django: :heavy_check_mark: 
- Flask: :heavy_check_mark:
- JAX-RS: :heavy_check_mark:
- Rails: :x: (would support `sort[]` though)
- PHP: :x: (would support `sort[]` though)
- ASP.NET: :heavy_check_mark:

So I'm going to say, let's change the parameter name to `sort[]`. I'll make this change with no objections in the next day or so.

Also, If we were to use Spring Data REST, it'd be possible to override the name of the parameter...

jewzaam replied:
Doe those frameworks guarantee order of the sort to match order in the request?

kahowell replied:
They guarantee ability to receive the query parameter in the order specified in the request (i.e. it is possible to read `sort[]` as an array or list). I didn't look into whether they understand "sort[]" natively, or how related libraries support sorting...

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+### Responses
+
+The response shall be a JSON object formatted according to Spring Data REST
+conventions.Specifically, the list is returned as a pluralized field name
```
Does this require any attribution to Spring? It's Apache 2.0 licensed so should be fine as long as we call it out assuming this sticks.

bvulaj replied:
This doesn't really have anything to do with Spring Data REST, they use the HAL spec http://stateless.co/hal_specification.html

bvulaj replied:
Sidenote, if we follow HAL conventions (Spring Data REST implements the spec), there are a number of clients, tools and frameworks we can use out there to help us.  Example: https://github.com/mikekelly/hal-browser

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+ - query for a specific subset of resources
+ - perform an action against a set of resources
+
+Additional functions shall be `POST` only, and parameters will be provided in
```
Do you think there is any value in distinguishing between idempotent PUT vs non-idempotent PATCH? I'm indifferent... Maybe leaning towards all POST all the time and calling it a day. But not sure.

kahowell replied:
For additional functions, not so much, as I think they're all essentially RPC-like, and I think it's reasonable to have them all be POST... My 2 cents.

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+Each function that can be performed against a specific resource must exist in
+two forms:
+ - non-batch
+ - batch
```
I think we should expect this of remote procedures against the collection too. That is, in findByLogin example, we should require batch support also like

``` json
[
  {"login": "bob"},
  {"login": "rob"}
]
```

alechenninger replied:
Thinking about bulk ops more, I actually like that lightblue's API (like OData) allows a bulk request with various operations against different resources, including basic CRUD ops. This has been a really useful optimization in practice.

Maybe this would be something we could elaborate on separately, since it doesn't exclude what you're proposing here of course.

Edit:
Here's OData's spec: http://www.odata.org/documentation/odata-version-3-0/batch-processing/

kahoali replied:
I am curious if we should support async operations for bulk requests.

alechenninger replied:
For all batch requests, we'll need clients to be aware there will be some limit on the amount of requests in the batch, and above this limit will return a client error (4xx) status code with descriptive message.

dcrissman replied:
> I am curious if we should support async operations for bulk requests.

From a lightblue perspective, only find can be performed async. All other operations are always sync in order to ensure order of operation.

bserdar replied:
That's changing: we are adding the concept of 'ordered' and 'unordered'
bulk requests. Unordered bulk requests will execute in parallel with no
particular order. Ordered bulk requests are processed in order, with only
consecutive find requests running in parallel.

On Fri, Jun 10, 2016 at 8:02 AM, Dennis Crissman notifications@github.com
wrote:

> In rest_specification/restful_service_specification.md
> https://github.com/lightblue-platform/lightblue-docs-specs/pull/43#discussion_r66617611
> :
> 
> > +the request body. The request body SHOULD be a JSON object with fields
> > +corresponding to parameter names.
> > +Examples:
> > +
> > +```json
> > +POST /user/findByLogin
> > +
> > +{
> > -  "login": "bob"
> >   +}
> >   +```
> >   +
> >   +Each function that can be performed against a specific resource must exist in
> >   +two forms:
> > - \- non-batch
> > - \- batch
> 
> I am curious if we should support async operations for bulk requests.
> 
> From a lightblue perspective, only find can be performed async. All other
> operations are always sync in order to ensure order of operation.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/lightblue-platform/lightblue-docs-specs/pull/43/files/19a07e2017455997f1b9c4bdb15edb6e3cceb8ce#r66617611,
> or mute the thread
> https://github.com/notifications/unsubscribe/ADgDDd8mAhU3Ls3YU53hwFBF73MjJk4Jks5qKW5xgaJpZM4IopeE
> .

alechenninger commented on rest_specification/restful_service_specification.md:
````diff
@@ -0,0 +1,240 @@
+  "status" : "success",
+  "data" : null
+}
+```
````
What about batch responses? Maybe response is array, with each response in same order as requests (so each index of a response aligns with the request at the same index).

kahowell replied:
My take on it is that the batch version would be an all-or-nothing, at least in terms of `status`.
How about the following?
- If the batch procedure is not all-or-nothing, and there is a error/failure then `data` contains a list of indices of arguments which had problems

alechenninger replied:
This one is a hard sell for me. What's the gain from all or nothing? It seems like it makes the spec more complicated so information can be consolidated. I'm not sure that is worth the tradeoff. Personally I'd rather see the simplicity of each request gets a response like any other response. Are there gains from all or nothing approach that I am missing?

kahoali replied:
I can see both sides of this coin and don't have a strong answer to weigh in with.

Instead, I'd like to ask a question:
What would we expect a recurring ( eg, periodic ) bulk request's logic to look like with a mixed-results scenario? Should they filter their responses line by line to see what failed and re-attempt?

How about for all-or-nothing?

I think all or nothing's client-side psuedocode is probably more attractive from a maintenance perspective but I am by no means sold on that being a compelling reason. Just my thoughts, at least.

dcrissman replied:
I am concerned that an error status on the "all-or-nothing" approach might indicate that the entire bulk operation was rolled back, and from a lightblue perspective that is not true.

alechenninger replied:
See Kevin's response [here](https://github.com/lightblue-platform/lightblue-docs-specs/pull/43/files/19a07e2017455997f1b9c4bdb15edb6e3cceb8ce#r67346228)

I'm not _crazy_ about that solution but I think I could learn to like it.

dcrissman replied:
What @kahowell suggests looks similar to me to what we do in lb today.

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+ - `status` field shall tell whether the function call was successful
+  - possible values: `success`, `fail`, `error`
+ - `data` shall have any feedback from the function call (function-specific)
+ - `message` shall contain a descriptive error message in the case of an error.
```
Sometimes there may be multiple errors. It might be nice if this could be an array. It might also be nice if this could be used for warnings also (such as when someone is using a deprecated code path for some reason or another). I've seen some API do this (I think it was Rally's) and found it especially useful when exploring the API to figure out what requests I needed to make.

kahowell replied:
This would deviate from the JSend standard though. I think we could still accomplish what you're getting at with something like:

``` javascript
{
  "status": "error"
  "message": "14 of 20 requests in the batch operation had errors",
  "data": {
    "size": 20,
    "errorCount": 14,
    "batch": [
      {
        "status": "error",
        "message": "failed to acquire resource"
      },
      // ...
    ]
  }
}
```

It's sort of recursive, but I quite like that it's possible to look at a batch call as just another function, while also being able to break it up into the individual operations from the batch. Thoughts?

dcrissman replied:
To be clear, I assume that successful requests would also be included, but with a different status?

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+| `POST`         | Create new resource(s) | -                             | JSON array   | -              |
+| `PUT`          | Replace resource(s)    | -                             | JSON array   | -              |
+| `PATCH`        | Update Resource(s)     | -                             | JSON array   | -              |
+| `GET`          | List resource(s)       | `fields`,`page`,`size`,`sort` | -            | paginated list |
```
How does fields param work? We have pretty deep objects where projections are commonly pretty complicated.

jewzaam replied:
Where is the query conditions?  I am assuming `fields` is for projection.

kahowell replied:
Yes fields is for projection... About fields... I'm tempted to say let's simplify by disallowing projection. We can handle large resources by splitting them out into sub-resources... Thoughts?

kahowell replied:
As to query conditions, my proposal is to never expose query conditions outside of a special function (such as the `findByLogin` example). I can make this explicit somehow if we agree on this.

bvulaj replied:
There's a possible middle-ground solution where we can provide pre-set projections rather than allowing whoever to project however they want.  Since we keep referencing Spring Data REST: http://docs.spring.io/spring-data/rest/docs/current/reference/html/#projections-excerpts

What does this look like? Rather than a `fields` param we have a `projection` param that just references a named, predefined projection.

Maybe this simplifies things without making them too simple?  Otherwise I agree with @kahowell

dcrissman replied:
> I'm tempted to say let's simplify by disallowing projection. 

My instinct is to be against this. We have seen performance gains in consumers of lightblue by being able to finely tune responses.

> There's a possible middle-ground solution where we can provide pre-set projections rather than allowing whoever to project however they want.

Does this become a management nightmare?
What about allowing multiple pre-set projections to be inclusively layered?

kahoali commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+
+Sorting shall be provided via the `sort` query parameter. `sort` may be
+specified more than once in order to sort by more than one field. `sort` values
+shall be of the form `sort={field}(,{direction})` where `direction` is `asc` or
```
Do we plan on having a default that's explicitly stated? I think it might be useful to state if unspecified the default would be <x>.

kahowell committed:
Update RESTful URI versioning specs

jewzaam commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,243 @@
+conventions:
+ - Responses must be JSON objects
+ - `status` field shall tell whether the function call was successful
+  - possible values: `success`, `fail`, `error`
```
With batches, is the expectation that all the requests in the batch have the same status?  This assumes transactions across all requests.  With lightblue we have status per request in bulk requests and partial success per request in cases that multiple things are being done, such as saving many documents.

kahowell replied:
See above comment w/ proposal.

jewzaam commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,243 @@
+
+Responses for functions shall follow [JSend](http://labs.omniti.com/labs/jsend)
+conventions:
+ - Responses must be JSON objects
```
With defaults, would be nice to return in the response what actual values were applied.  So if paging has default size and you don't specify a size, return in the response the size used.  Same with sorting if there's a default direction.

kahowell replied:
My proposal would be that we don't do paging for batches function calls at all, only for resources... We can set a limit per-function on the batch size allowed. Open to suggestions on default limits.

BVulaj commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,243 @@
+    ]
+  },
+  "page": {
+    "size": 20,
```
It's worth pointing out that this page section is not part of HAL, but just part of Spring Data REST's `Page` section of the response.  I only bring this up because the operations behind finding this information can be expensive depending on the persistence implementation.  For example, a `count *` in PostgreSQL is *really* slow, which is effectively what gets executed on each page to get this info.  I'm a larger fan of adhering to something more like Spring's `Slice` object, which only exposes whether or not there are more pages, with prev / next links.

bvulaj commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,243 @@
+      ...
+    ]
+  },
+  "page": {
```
It's worth pointing out that this page section is not part of HAL, but just part of Spring Data REST's `Page` portion of the response.  I only bring this up because the operations behind finding this information can be expensive depending on the persistence implementation.  For example, a `count *` in PostgreSQL is _really_ slow, which is effectively what gets executed on each page to get this info.  I'm a larger fan of adhering to something more like Spring's `Slice` object, which only exposes whether or not there are more pages, with prev / next links.

alechenninger replied:
Just documented where we landed in last meeting here: I think it was that not all of the page metadata should be required, only the current page number. That way, if you can do a fast total elements count, provide it. If not, don't.

That does mean though that users could not rely on it being there without consulting documentation, and removing them would be a breaking change. Could be another opportunity for auto-generated clients / documentation to smooth over.

jewzaam commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,243 @@
+| findByLogin | find user(s) by login     |
+
+As illustrated, links may be for related entities, pagination, relevant
+functions, etc.
```
Please add section(s) for authentication.

jewzaam replied:
Please add some complex examples.  One thing I can think of is insert new legal entity and user with default paging and sort response by login with default order.

kahowell committed:
Change to `sort[]`, use javascript for examples

Using javascript instead of JSON allows us to use comments without
getting syntax errors :-)

kahowell committed:
Add authentication/authorization to rest specs

alechenninger commented:
Hi all, I opened up a PR against this branch to propose query params for resource operations. Please rip it to shreds!

https://github.com/kahowell/lightblue-docs-specs/pull/1

alechenninger commented on rest_specification/restful_service_specification.md:
```diff
@@ -0,0 +1,240 @@
+GET /user
+
+{
+  "_embedded": {
```
Two questions WRT responses:

**1: What about error responses for CRUD operations?**
There are some standards here:
https://github.com/blongden/vnd.error
https://datatracker.ietf.org/doc/rfc7807/?include_text=1

We could also consider just using JSend error responses that we use for RPC calls below.

**2: Why have different responses for CRUD operations vs additional functions?**
I'd like to see us accommodate both cases with one standard. Ideally, all successes look consistent and all errors look consistent. To me they are not different enough to warrant having to understand how to parse different responses. Again, maybe not so bad for those of us comfortable with JSON, HTTP, HAL, etc, but trying to empathize with other users. And still, it would be nice for developer sanity, too, I think (imagine trying to overlay multiple frameworks / libraries for different standards vs just using one for one standard). Plus, it shouldn't be so hard to consolidate, they are effectively just responses with data either way. This is consistent with most examples I've seen in the wild, too (they all use the same response envelope / metadata).

So, why not use HAL for additional function responses, too? Or JSend for both. I don't really care what standard, just that we're not arbitrarily inconsistent. What do you guys think?

bvulaj replied:
I agree.  As long as we're consistent in what we do, I can be made happy.  HAL has a lot of traction from what I've seen, so I will vote for that, but consistency is king.

If anyone finds it useful, I can put up a basic Spring Data REST application to toy around with.  It would help anyone who hasn't toyed with HAL get a feel for it.

dcrissman commented:
Moving discussion to https://github.com/dcrissman/restful-services-spec

Discussion points will be created as issues so we can better track our progress.

Closing this as a PR
